RIOT-OS/RIOT
# arduino/sketches: build sketches as a module

Generate a module for arduino sketches in a subfolder of BINDIR.
This prevents issues when doing concurrent builds or out of tree build with readonly sources.

This fixes #5848

The concept is instead of generating a cpp file in the source directory, it is generated into the build directory. However, this requires defining a new module for this directory.
It is now generated during build and not parsing.
I did not addressed the "write atomically" as I think it is not necessary anymore.

To do this, I replaced the previous bash script by a Makefile script which creates the module as `link` target prerequisite.


### Testing procedure

As it was not asked before, I will give a testing procedure:

Setup:

* Clean
*    create bin
*    remove write permission in the application directory

```
git clean -xdf examples/arduino_hello-world
mkdir -p examples/arduino_hello-world/bin
chmod -w examples/arduino_hello-world
make -C examples/arduino_hello-world clean all BOARD=arduino-mega2560
```

Cleanup

* `chmod u+w examples/arduino_hello-world`

It works with this PR.

It fails on master, as we try to write in the directory and so the main function is never declared.
You can also notice that the Permission denied are there before even starting to build. make clean also tries to create the file.

Fixes #5848
